driver/main.c

```diff
@@ -90,31 +346,81 @@ int main(int argc, char *argv[]) {
     // Currently, these datasets need to be opened before calling
     // RDySetup to avoid an error. Possible issue for the error could
     // be that RDySetup is setting a default DM for all VecLoads.
-    PetscBool rain_specified, bc_specified;
-    char      rainfile[PETSC_MAX_PATH_LEN], bcfile[PETSC_MAX_PATH_LEN];
-    PetscCall(PetscOptionsGetString(NULL, NULL, "-rain", rainfile, sizeof(rainfile), &rain_specified));
-    PetscCall(PetscOptionsGetString(NULL, NULL, "-bc", bcfile, sizeof(bcfile), &bc_specified));
+    PetscBool flag, bc_specified;
 
-    PetscBool interpolate_rain = PETSC_FALSE, interpolate_bc = PETSC_FALSE;
-    PetscCall(PetscOptionsGetBool(NULL, NULL, "-interpolate_rain", &interpolate_rain, NULL));
-    PetscCall(PetscOptionsGetBool(NULL, NULL, "-interpolate_bc", &interpolate_bc, NULL));
+    Rain rain_dataset;
+
+    // set default rainfall
+    rain_dataset.type = CONSTANT;
 
-    Vec          rain_vec = NULL, bc_vec = NULL;
-    PetscScalar *rain_ptr = NULL, *bc_ptr = NULL;
-    PetscInt     nrain, nbc;
+    char bcfile[PETSC_MAX_PATH_LEN];
 
-    if (rain_specified) {
-      PetscCall(OpenData(rainfile, &rain_vec, &nrain));
-      PetscCall(VecGetArray(rain_vec, &rain_ptr));
+    PetscCall(
+        PetscOptionsGetString(NULL, NULL, "-homogeneous_rain", rain_dataset.homogeneous.filename, sizeof(rain_dataset.homogeneous.filename), &flag));
+    if (flag) {
+      rain_dataset.type = HOMOGENEOUS;
+    }
+    PetscCall(PetscOptionsGetBool(NULL, NULL, "-interpolate_spatially_homogeneous_rain", &rain_dataset.homogeneous.temporally_interpolate, NULL));
+
+    PetscBool sp_hetero_dir_flag;
+    PetscCall(PetscOptionsGetString(NULL, NULL, "-heterogeneous_rain_dir", rain_dataset.heterogeneous.dir, sizeof(rain_dataset.heterogeneous.dir),
+                                    &sp_hetero_dir_flag));
+    PetscInt nvalues = 5;
+    PetscInt date[nvalues];
+    PetscInt ndate = nvalues;
+    PetscCall(PetscOptionsGetIntArray(NULL, NULL, "-heterogeneous_rain_start_date", date, &ndate, &flag));
+    if (flag) {
+      PetscCheck(ndate == nvalues, PETSC_COMM_WORLD, PETSC_ERR_USER, "Expect 5 values when using -heterogeneous_rain_start_date YY,MO,DD,HH,MM");
+      PetscCheck(rain_dataset.type != HOMOGENEOUS, PETSC_COMM_WORLD, PETSC_ERR_USER,
+                 "Can only specify homogeneous or heterogeneous rainfall datasets.");
+      PetscCheck(sp_hetero_dir_flag == PETSC_TRUE, PETSC_COMM_WORLD, PETSC_ERR_USER,
+                 "Need to specify path to spatially heterogeneous rainfall via -heterogeneous_rain_dir <dir>");
+
+      rain_dataset.type = HETEROGENEOUS;
```

Comment:
One last tiny comment, just because I think the syntax for initializing C structs is cool:

Lines 379-393 could be replaced with the following, which automatically zeroes out all fields that aren't initialized:

```
rain_dataset = (Rain)({
  .type = HETEROGENEOUS,
  .heterogeneous.start_date = {
    .tm_year = date[0] - 1900,
    .tm_mon = date[1] - 1;
    .tm_mday = date[2],
    .tm_hour = date[3],
    .tm_min = date[4],
  },
  .heterogeneous.current_date = {
    .tm_year = date[0] - 1900,
    .tm_mon = date[1] - 1,
    .tm_mday = date[2],
    .tm_hour = date[3],
    .tm_min = date[4],  // note: the trailing comma is acceptible here, making it easy to add more
};
```

If you initialize the variable when you declare it, you don't need the `(Rain)` cast. Then it's just

```
Rain rain_dataset = {
  .type = HETEROGENEOUS,
  ...
};
```